fem/bilininteg_hcurl.cpp

```diff
@@ -1322,12 +1633,325 @@ static void PACurlCurlApply3D(const int D1D,
+
+                           for (int dy = 0; dy < D1Dy; ++dy)
+                           {
+                              const double wy = Bo[dy][qy];
+
+                              for (int dx = 0; dx < D1Dx; ++dx)
+                              {
+                                 const double t = X[dz][dy][dx];
+                                 const double wx = t * Bc[dx][qx];
+                                 const double wDx = t * Gc[dx][qx];
+
+                                 u += wDx * wy * wz;
+                                 v += wx * wy * wDz;
+                              }
+                           }
+                        }
+
+                        curl[qy][qx][0] -= v; // -(u_1)_{x_2}
+                        curl[qy][qx][2] += u; // (u_1)_{x_0}
+                     }
+                     else // z component
+                     {
+                        // \hat{\nabla}\times\hat{u} is [(u_2)_{x_1}, -(u_2)_{x_0}, 0]
+
+                        for (int dz = 0; dz < D1Dz; ++dz)
+                        {
+                           const double wz = Bo[dz][qz];
+
+                           for (int dy = 0; dy < D1Dy; ++dy)
+                           {
+                              const double wy = Bc[dy][qy];
+                              const double wDy = Gc[dy][qy];
+
+                              for (int dx = 0; dx < D1Dx; ++dx)
+                              {
+                                 const double t = X[dz][dy][dx];
+                                 const double wx = t * Bc[dx][qx];
+                                 const double wDx = t * Gc[dx][qx];
+
+                                 u += wDx * wy * wz;
+                                 v += wx * wDy * wz;
+                              }
+                           }
+                        }
+
+                        curl[qy][qx][0] += v; // (u_2)_{x_1}
+                        curl[qy][qx][1] -= u; // -(u_2)_{x_0}
+                     }
+                  } // qx
+               } // qy
+            } // tidz == qz
+
+            osc += D1Dx * D1Dy * D1Dz;
+            MFEM_SYNC_THREAD;
+         } // c
+
+         double dxyz1 = 0.0;
+         double dxyz2 = 0.0;
+         double dxyz3 = 0.0;
+
+         MFEM_FOREACH_THREAD(dz,z,D1D)
+         {
+            const double wcz = Bc[dz][qz];
+            const double wcDz = Gc[dz][qz];
+            const double wz = (dz < D1D-1) ? Bo[dz][qz] : 0.0;
+
+            MFEM_FOREACH_THREAD(dy,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(dx,x,D1D)
+               {
+                  for (int qy = 0; qy < Q1D; ++qy)
+                  {
+                     const double wcy = Bc[dy][qy];
+                     const double wcDy = Gc[dy][qy];
+                     const double wy = (dy < D1D-1) ? Bo[dy][qy] : 0.0;
+
+                     for (int qx = 0; qx < Q1D; ++qx)
+                     {
+                        const double O11 = sop[0 + (6*qx) + (6*Q1D*qy)];
```

Comment:
These offset computation could be avoided by some DeviceTensor that can be used inside the kernel body.